File: windows/src/Elevate.Core/Managed/Branding.cs
namespace Elevate.Core.Managed;

/// <summary>
/// The organization's co-branding, resolved from <see cref="ManagedConfiguration"/> into the exact
/// strings each surface renders. Resolving to null is the unbranded case: every caller renders
/// exactly what it rendered before this type existed, so an unbranded install is unaffected.
///
/// Elevate's own name is never replaced — the organization's name is added beside it. Port of the
/// Swift <c>Branding</c> struct; see
/// <c>docs/superpowers/specs/2026-09-18-enterprise-cobranding-design.md</c>.
/// </summary>
public sealed record Branding(
    string OrganizationName,
    OrganizationTitleStyle TitleStyle,
    Uri? SupportUrl,
    string? SupportEmail)
{
    /// <summary>The panel header's second line, under "Elevate". Null for the None style.</summary>
    public string? HeaderCaption => TitleStyle switch
    {
        OrganizationTitleStyle.By => $"by {OrganizationName}",
        OrganizationTitleStyle.ManagedBy => $"Managed by {OrganizationName}",
        _ => null,
    };

    public string FirstRunLine => $"Provided by {OrganizationName}.";

    public string SupportLabel => $"{OrganizationName} IT";

    public bool HasSupport => SupportUrl is not null || SupportEmail is not null;

    /// <summary>
    /// The single link a "Get help" control opens: the help desk URL, or the address as a
    /// <c>mailto:</c>. One place on purpose, so no view builds a <c>mailto:</c> of its own.
    /// </summary>
    public Uri? SupportDestination =>
        SupportUrl ?? (SupportEmail is { } email ? new Uri($"mailto:{email}") : null);

    /// <summary>
    /// The one-line contact appended to CLI failures. Prefers the URL: a help desk portal lists the
    /// address, but an address does not list the portal.
    /// </summary>
    public string? SupportLine =>
        (SupportUrl?.ToString() ?? SupportEmail) is { } target
            ? $"Need help? {SupportLabel} — {target}"
            : null;

    /// <summary>One line for the diagnostics report: the name, its phrasing, and the help desk.</summary>
    public string DiagnosticsLine
    {
        get
        {
            var parts = new List<string> { $"{OrganizationName} ({TitleStyle.Name()})" };
            if (SupportUrl is not null)
            {
                parts.Add(SupportUrl.ToString());
            }

            if (SupportEmail is not null)
            {
                parts.Add(SupportEmail);
            }

            return string.Join(" · ", parts);
        }
    }

    public static Branding? Resolve(ManagedConfiguration config)
    {
        ArgumentNullException.ThrowIfNull(config);

        return config.OrganizationName is not { } name
            ? null
            : new Branding(
                name,
                config.OrganizationTitleStyle ?? OrganizationTitleStyle.By,
                config.OrganizationSupportUrl,
                config.OrganizationSupportEmail);
    }
}

File: macos/Sources/ElevateCore/Managed/Branding.swift
import Foundation

/// The organization's co-branding, resolved from `ManagedConfiguration` into the exact strings each
/// surface renders. Resolving to `nil` is the unbranded case: every caller renders exactly what it
/// rendered before this type existed, so an unbranded install is unaffected by co-branding entirely.
///
/// Elevate's own name is never replaced — the organization's name is added beside it. See
/// `docs/superpowers/specs/2026-09-18-enterprise-cobranding-design.md` §1.
public struct Branding: Hashable, Sendable {
    public let organizationName: String
    public let titleStyle: OrganizationTitleStyle
    public let supportUrl: URL?
    public let supportEmail: String?

    /// The panel header's second line, under "Elevate". `nil` for the `none` style.
    public var headerCaption: String? {
        switch titleStyle {
        case .by: "by \(organizationName)"
        case .managedBy: "Managed by \(organizationName)"
        case .none: nil
        }
    }

    public var firstRunLine: String { "Provided by \(organizationName)." }

    public var supportLabel: String { "\(organizationName) IT" }

    public var hasSupport: Bool { supportUrl != nil || supportEmail != nil }

    /// The single link a "Get help" control opens: the help desk URL, or the address as `mailto:`.
    /// One place on purpose, so no view builds a `mailto:` of its own.
    public var supportDestination: URL? {
        if let supportUrl { return supportUrl }
        guard let supportEmail else { return nil }
        return URL(string: "mailto:\(supportEmail)")
    }

    /// The one-line contact appended to CLI failures. Prefers the URL: a help desk portal lists the
    /// address, but an address does not list the portal.
    public var supportLine: String? {
        guard let target = supportUrl?.absoluteString ?? supportEmail else { return nil }
        return "Need help? \(supportLabel) — \(target)"
    }

    /// One line for the diagnostics report: the name, how it is phrased, and the help desk.
    public var diagnosticsLine: String {
        var parts = ["\(organizationName) (\(titleStyle.rawValue))"]
        if let supportUrl { parts.append(supportUrl.absoluteString) }
        if let supportEmail { parts.append(supportEmail) }
        return parts.joined(separator: " · ")
    }

    public static func resolve(from config: ManagedConfiguration) -> Branding? {
        guard let name = config.organizationName else { return nil }
        return Branding(
            organizationName: name,
            titleStyle: config.organizationTitleStyle ?? .by,
            supportUrl: config.organizationSupportUrl,
            supportEmail: config.organizationSupportEmail
        )
    }
}
